### ISSUE TITLE
Scripts - scriptable lifecycle events

### ISSUE THREAD

**Issue by mthenw** [serverless/serverless#4683]
# This is a Feature Proposal

## Description

Serverless Framework users should be able to define scripts (shell commands) that run on different phases (lifecycle events) so they are able to customize their workflows without creating plugins. It becomes more important with the introduction of new compiled runtimes (Golang, Java, C#) but even for Node.js that would be helpful for running tools like babel or webpack. Right now some of that use-cases are solved by very specific and opinionated plugins (webpack, etc.) or by general plugins like [1] and [2]. It clearly indicates that this feature is needed, and helpful. It could be also used for running automated tests and other verification steps before deployment without using external tools/plugins.

[1] https://github.com/uswitch/serverless-hooks-plugin
[2] https://github.com/weixu365/serverless-scriptable-plugin

### What is the use case that should be solved. The more detail you describe this in the easier it is to understand for us.

Use cases:

- compiling function (Golang, C#, Java)
- running minimization tools and transpilers
- running automated tests

### If there is additional config how would it look

New root level filed called `scripts` should be introduced
```yaml
...

scripts:
  package:initialize: "GOOS=linux go build -o bin/hello main.go" 

...
```

In the first pass it should allow running one script for one lifecycle hook.

**Comment by HyperBrain** [comment]
Hi @mthenw . Please have a look at https://www.npmjs.com/package/serverless-plugin-scripts. It offers exactly what you describe here. You can inject a script execution into any lifecycle event, the same as you could do with a plugin.

**Comment by mthenw** [comment]
Hey @HyperBrain.

I'm aware of this plugin and other plugins that provide similar functionality. I believe it should be supported by the framework because of the following reasons.

1. Most of the runtimes (Golang, C#, Java) **require** running some kind of compilation step before deployment (this is also true for more complicated nodejs and python service). Lack of this functionality means that user has to fallback to other tools for this very basic use-case.

2. Being forced to use a plugin for this functionality means that my development workflow is more complicated because I need to use npm for installing a plugin. Of course, npm is needed for installing the framework but storing and maintaining package.json only because of that is IMHO unnecessary step.

**Comment by horike37** [comment]
+1 :+1: 
I agree with this proposal.

**Comment by HyperBrain** [comment]
I also agree, but we should check if such an integration into core would introduce security issues. Having a general script execution ability in there, would let anyone execute scripts on the build server just by adding them into the `serverless.yml`. As soon as the requirements and how it will be done is clear, it should be evaluated in regards of security.

**Comment by softprops** [comment]
+1
